skaffold/src/test/kotlin/com/google/container/tools/skaffold/run/BaseSkaffoldSettingsEditorTest.kt

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright 2018 Google Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.container.tools.skaffold.run
+
+import com.google.common.truth.Truth.assertThat
+import com.google.container.tools.skaffold.SkaffoldFileService
+import com.google.container.tools.test.ContainerToolsRule
+import com.google.container.tools.test.TestService
+import com.google.container.tools.test.UiTest
+import com.google.container.tools.test.expectThrows
+import com.intellij.mock.MockVirtualFile
+import com.intellij.openapi.options.ConfigurationException
+import com.intellij.testFramework.EdtTestUtil
+import com.intellij.util.ThrowableRunnable
+import io.mockk.every
+import io.mockk.impl.annotations.MockK
+import io.mockk.impl.annotations.SpyK
+import org.junit.Before
+import org.junit.Rule
+import org.junit.Test
+
+/**
+ * Unit tests for base Skaffold settings support defined in [BaseSkaffoldSettingsEditor] and
+ * [BaseSkaffoldRunConfiguration].
+ */
+class BaseSkaffoldSettingsEditorTest {
+    @get:Rule
+    val containerToolsRule = ContainerToolsRule(this)
+
+    @SpyK
+    @TestService
+    private var mockSkaffoldFileService: SkaffoldFileService = SkaffoldFileService()
+
+    @MockK
+    private lateinit var mockSkaffoldSettings: BaseSkaffoldRunConfiguration
+
+    private lateinit var baseSkaffoldSettingsEditor: BaseSkaffoldSettingsEditor
+
+    @Before
+    fun setUp() {
+        EdtTestUtil.runInEdtAndWait(ThrowableRunnable {
+            baseSkaffoldSettingsEditor = BaseSkaffoldSettingsEditor()
+            // calls getComponent() to initialize UI first, IDE flow.
+            baseSkaffoldSettingsEditor.component
+        })
+
+        // default is project with no Skaffold files
+        every { mockSkaffoldFileService.findSkaffoldFiles(any()) } returns listOf()
+        // return test fixture project for settings
+        every { mockSkaffoldSettings.project } answers
+            { containerToolsRule.ideaProjectTestFixture.project }
+    }
+
+    @Test
+    @UiTest
+    fun `settings are invalid for project with no existing Skaffold files`() {
```

Comment:
regarding these freeform function names: I like this is a lot, but maybe lets still give it some structure so that we know exactly what to expect from the tests when reading it:

i.e. in Java we did something like 

`resetFrom_whenNoSkaffoldFilesFound_throwsConfigurationException`

This makes it very clear the behavior and expectation.

For Kotlin, if we structured it similarly we could have something like (feel free to deviate from this):

`when project has no Skaffold files a ConfigurationException is thrown`

wdyt?

Reply:
Sounds like a good idea, with this freeform names I sometimes deviate a bit, this simple rule should help having clearer names. Excellent suggestion!